parlai/scripts/multiprocessing_train.py

```diff
@@ -62,50 +64,60 @@ def multiprocess_train(rank, opt, port=61337, gpu=None, hostname='localhost'):
 
     # Suppress output of workers except the main host.
     if opt.get('verbose') or rank != 0:
-        print_prefix = '[rank:{:2d}]'.format(rank)
+        print_prefix = '[rank:{:3d}]'.format(rank)
     else:
         print_prefix = None
-    distributed_utils.override_print(
-        suppress=(not opt.get('verbose') and rank != 0), prefix=print_prefix
-    )
-
-    # perform distributed setup, ensuring all hosts are ready
-    torch.cuda.set_device(opt['gpu'])
-    dist.init_process_group(
-        backend="nccl",
-        init_method="tcp://{}:{}".format(hostname, port),
-        world_size=opt['distributed_world_size'],
-        rank=rank,
-    )
-    print("Distributed group initialized")
+    suppress_output = not opt.get('verbose') and rank != 0
 
-    # Run the actual training
-    return single_train.TrainLoop(opt).train()
+    with distributed_utils.override_print(suppress_output, print_prefix):
+        # perform distributed setup, ensuring all hosts are ready
+        torch.cuda.set_device(opt['gpu'])
+        dist.init_process_group(
+            backend="nccl",
+            init_method="tcp://{}:{}".format(hostname, port),
+            world_size=opt['distributed_world_size'],
+            rank=rank,
+        )
+        print("Distributed group initialized")
 
+        # Run the actual training
+        return single_train.TrainLoop(opt).train()
 
-def main():
-    parser = single_train.setup_args()
-    parser.add_distributed_training_args()
-    parser.set_defaults(distributed_world_size=torch.cuda.device_count())
-    opt = parser.parse_args()
-
-    port = random.randint(32000, 48000)
 
+def launch_and_train(opt, port):
+    """Perform a fork() to many processes."""
     # Launch multiple subprocesses
     spawncontext = torch.multiprocessing.spawn(
         multiprocess_train,
         (opt, port),
-        nprocs=opt['distributed_world_size'],
+        nprocs=opt['distributed_world_size'] - 1,
         join=False,
     )
 
     try:
+        # rank is offset by -1 to make the root process be rank 0
```

Comment:
can you elaborate a little bit why you offset by -1 here and then offset by +1 above?

Reply:
Sure. It should be added as a comment then.

the calls to spawn() always start at 0, and that's nonconfigurable. However, we want rank 0 to be the main thread, so that we can successfully capture the response and return the retval properly.